samples/rpi/blinky.cpp

```diff
@@ -0,0 +1,132 @@
+#include <chrono>
+#include <iostream>
+#include <string>
+
+#include <Wink/address.h>
+#include <Wink/machine.h>
+#include <Wink/state.h>
+
+int main(int argc, char **argv) {
+  if (argc < 3) {
+    error() << "Incorrect parameters, expected <spawner> <address>\n"
+            << std::flush;
+    return -1;
+  }
+
+  Address spawner(argv[1]);
+  Address address(argv[2]);
+  UDPSocket socket;
+  Machine m(spawner, address, "rpi/Blinky", socket);
+
+  const auto reader = "rpi/GPIOReader";
+  const auto writer = "rpi/GPIOWriter";
+
+  const auto buttonPin = 10;
+  const auto led1Pin = 11;
+  const auto led2Pin = 12;
+
+  auto led1on = false;
+
+  std::map<const std::string, const Address &> addresses;
+
+  m.AddState(std::make_unique<State>(
+      // State Name
+      "main",
+      // Parent State
+      "",
+      // On Entry Action
+      [&]() {
+        m.Spawn(reader);
+        m.Spawn(writer);
+        m.GotoState("loop");
```

Comment:
Rather than having a GotoState here, I'd suggest when you construct `Machine m()` you give the name of the initial state in the constructor. Then when the machine starts running your infrastructure will enter the appropriate states when the first message arrives. Maybe its just a style difference, but in my SMs I perform the transition after leaving the "Receiver/ProcessMessage" function. If you process GotoState immediately what happens when you execute a a GotoState in the middle of a function call, so what happens if the order was:
```
  m.GotoState("loop");
  m.Spawn(reader);
  m.Spawn(writer);
```

In my current Rust SM I transition by returning the target in from the "Receiver/ProcessMsg" function, that way the the user knows that the transition is after processing the message. Not only that, but if the message was "not handled" then the parents can transition if the child didn't. Note; I use "first" transition wins if there are multiple transitions in a child parent set but that could change, someone has to "win" if there are multiple I chose first as the winder as the first done by a "Receiver/ProcessMsg" that is "more" specialized, but that's my rationale.

Side note, I'm probably going to go back to using a "GotoState" function moving forward, which is what I had in Java and C before that. The reason is in Rust I ended up needing an extra level of indirection by returning it, but we'll see.

Reply:
> Rather than having a GotoState here, I'd suggest when you construct Machine m() you give the name of the initial state in the constructor. Then when the machine starts running your infrastructure will enter the appropriate states when the first message arrives.

By default the SM starts in the first state you added to it, but you can give an explicit initial state to `Start` if you wish.

> If you process GotoState immediately what happens when you execute a a GotoState in the middle of a function call, so what happens if the order was:

In that situation, the children would get spawned after the `main` state had exited and the `loop` state was entered, when control flow had returned out of `GotoState`.

Good point about when things happen, right now it is all immediate, will need to think more on this. Especially multiple `GotoState` calls, currently every transition is fulfilled, maybe it makes sense, or maybe it is better that `GotoState` doesn't return and instead the machine stops the current action and then transitions, again need more thought here.

Reply:
What I'd suggest is that GotoState would only set a new instance variable, Machine::next_state, and then upon returning from the receiver Machine::handleMessage would invoke the onExit method(s). And then upon receiving the next message Machine::handleMessage invokes the onEnter method(s) before invoking the receiver(s).

But now that there can be multiple invocations of GotoState(dest_state) without an action, this means you need to decide which of the multiple dest_states to do use. I've chosen that the "first" invocation is the one that wins. I decided that because my opinion is the first one is more "specific" so it should win.

Reply:
> What I'd suggest is that GotoState would only set a new instance variable, Machine::next_state, and then upon returning from the receiver Machine::handleMessage would invoke the onExit method(s). And then upon receiving the next message Machine::handleMessage invokes the onEnter method(s) before invoking the receiver(s).

I think that makes sense. The only thing I would change is the last bit; I don't think it makes sense to wait until the next message before executing the `OnEntry`. Instead `GotoState` would set an instance variable as you suggested, then when the action (OnExit, OnEntry, Receiver) returns, the current state's OnExit is called, followed by the next state's OnEntry, and then it awaits the next message.

> But now that there can be multiple invocations of GotoState(dest_state) without an action, this means you need to decide which of the multiple dest_states to do use. I've chosen that the "first" invocation is the one that wins. I decided that because my opinion is the first one is more "specific" so it should win.

Yeah I'm not sure about this, on the one hand I like the simplicity of "first wins" but not so much the implicitly of "any other GotoState calls are ignored". On the other hand I can see the situation where a handler calls GotoState to set a "default" destination state, then as it is parsing the rest of the message, it might override that destination with another state.

Reply:
Created https://github.com/AletheiaWareLLC/Wink/issues/47 to track